### DISCUSSION TITLE
Supporting Timespan math operations

### DISCUSSION THREAD

**Discussion by AyhamAl-Ali** [SkriptLang/Skript#4720] [Suggestions]
💡 Would you like to support Timespan in math operations? see example below

## How it looks
:small_blue_diamond:  **What to be able to do**
```v
send "%10 seconds - difference between {_t1} and {_t2}%"
```

:small_orange_diamond: **Instead of**
```v
set {_diff} to 10 seconds
remove difference between {_t1} and {_t2} from {_diff}
send "Time remaining: %{_diff}%"
```

More examples:
```v
# subtraction
send "%20 seconds - 5 seconds%"

# addition
send "%time from 10 seconds and {_t1} and {_t2}%"
```
---

## :small_blue_diamond: Proposal 01

Supporting this in [ExprArithmetic](https://docs.skriptlang.org/expressions.html#ExprArithmetic), this will allow all kind of math operations between Timespans such as `/ * - + ^`

## :small_blue_diamond: Proposal 02

Only supporting addition and subtraction, probably by making a new expression for this feature

## :small_blue_diamond: Proposal 03

Making it english-like expression, something like 
- `time[span] left of %timespan% from %timespans%` - subtraction
- `time[span] from %timespan% and %timespans%` - addition

**Comment by Ankoki** [comment]
I like proposal 1. You could possibly just take in %object% for the ExprArithmetic patterns. From there you could just check if its instanceof Number or Timespan, which will make it easier to add support for things later. The only downside I see to this is that you won't get any error at init for `"hey" - "ur mum"`. Possibly use %number/timespan%? Then do the maths with Timespan#getTicks_i() and return that :D this is a great idea though.

**Reply by AyhamAl-Ali** [reply]
:+1: 
You can check instanceof of expressions in init and error if not Number or Timespan but I had a better idea to create a custom type like `%arithmeticable%` which would allow all what we decide to be arithmeticable

**Reply by Ankoki** [reply]
I like the `%arithmeticable%` idea. I don't think checking instanceof in init is reliable though, especially in terms of variables as they can be changed at runtime.

**Comment by Mwexim** [comment]
Proposal 1 is by far the most suitable. That said, multiplication leaves some discussion as a timespan technically has a unit attached to it, unlike numbers.

There should, however, be a general implementation that supports **all** number-like types. This means that **any type that has an arithmetic operations registered and uses itself as the change type, should automatically have this functionality**.
- Date has an Arithmetic instance registered but uses Timespan as the changer and is therefore not suitable for this system.

This way all addon developers can hook into this system. Another huge advantage is that the vector math expression can all be reduced and added to this system!

**Reply by AyhamAl-Ali** [reply]
Indeed, this applies to the previous [comment](https://github.com/SkriptLang/Skript/discussions/4720#discussioncomment-2533934) adding a custom type like `%arithmeticable%` would make this a lot easier for any kind of implementation later on

**Comment by bluelhf** [comment]
I'm not familiar with ExprArithmetic or timespans, but couldn't we just make timespans implement Number, returning their tick values as needed?

**Reply by AyhamAl-Ali** [reply]
We could've added a converter from `Number to Timespan` and vise versa but the problem here it will return a Number type which is unexpected

**Comment by TPGamesNL** [comment]
IMO the implementation should be done by making ExprArithmetic take %object%, then check at runtime if the return type of the given expression has arithmetic (or is Object.class). We should also extend the Arithmetic class by adding a method like acceptsOperation to check if it e.g. supports multiplication, division and exponentiation.
